### DISCUSSION TITLE
Expected behavior of getsockname for unbound socket

### DISCUSSION THREAD

**Discussion by metamon123** [ANLAB-KAIST/KENSv3#76] [Project Q&A 2. Transport Layer (TCP)] [answered]
Hi. I wonder the expected behavior of syscall `getsockname` for unbound socket in KENS.

In the following references which are mentioned in PA#2 doc, there is no well-defined specification for unbound socket input.
* https://linux.die.net/man/2/getsockname
  * no specification for unbound socket
* https://linux.die.net/man/3/getsockname
  * _If the socket has not been bound to a local name, the value stored in the object pointed to by address is unspecified._

Specifically, I wonder these three.
1. What should `getsockname` return for unbound socket input, considering future PA3 too? Success (0) vs Error (-1)
2. What should `getsockname` set on the given addr input for unbound socket input? Can it just skip setting addr?
3. If connection is established without user's `bind` call, should `getsockname` return the address that system (i.e. KENS) arbitrarily assigned?
    - ex. `connect` without `bind` / `listen` without `bind` (I'm not sure whether latter is legitimate or not)

**Comment by Pusnow** [comment, marked as answer]
> 1. What should getsockname return for unbound socket input, considering future PA3 too? Success (0) vs Error (-1)
> 2. What should getsockname set on the given addr input for unbound socket input? Can it just skip setting addr?

Unbound socket would return `INADDR_ANY` (0.0.0.0) and zero for return value, and the result will not used in the PA3.


> 3. If connection is established without user's bind call, should getsockname return the address that system (i.e. KENS) arbitrarily assigned?
 
For the connected sockets, an address in 5-tuple is returned.


If you're curious about the specific behavior of a system call, you can try it on a Linux machine.
